File: app/api/moderation_violations.php
<?php
/********************************************
 * Moderation Violations API
 * Get pending violations and take actions
 ********************************************/

header('Content-Type: application/json');

require_once __DIR__ . '/../database/Database.php';
require_once __DIR__ . '/../includes/ModerationNotifier.php';

$db = Database::getInstance();
$notifier = new ModerationNotifier();

try {
    $action = $_GET['action'] ?? $_POST['action'] ?? 'list';

    switch ($action) {
        case 'list':
            // Get all pending violations
            $status = $_GET['status'] ?? 'pending';

            $violations = $db->query("
                SELECT
                    v.*,
                    a.title as ad_title,
                    a.description as ad_description,
                    a.category_slug,
                    c.company_name,
                    c.email as company_email,
                    cat.category_name
                FROM moderation_violations v
                LEFT JOIN ads a ON v.ad_id = a.ad_id
                LEFT JOIN companies c ON v.company_slug = c.company_slug
                LEFT JOIN categories cat ON a.category_slug = cat.category_slug
                WHERE v.status = ?
                ORDER BY v.severity DESC, v.created_at DESC
                LIMIT 50
            ", [$status]);

            // Parse violations JSON
            foreach ($violations as &$v) {
                $v['violations_data'] = json_decode($v['violations'], true);
            }

            echo json_encode([
                'success' => true,
                'violations' => $violations,
                'count' => count($violations)
            ], JSON_PRETTY_PRINT);
            break;

        case 'stats':
            // Get violation statistics
            $stats = $db->queryOne("
                SELECT
                    COUNT(*) as total,
                    SUM(CASE WHEN status = 'pending' THEN 1 ELSE 0 END) as pending,
                    SUM(CASE WHEN status = 'resolved' THEN 1 ELSE 0 END) as resolved,
                    SUM(CASE WHEN severity = 4 THEN 1 ELSE 0 END) as critical,
                    SUM(CASE WHEN severity = 3 THEN 1 ELSE 0 END) as high,
                    SUM(CASE WHEN severity = 2 THEN 1 ELSE 0 END) as medium,
                    SUM(CASE WHEN severity = 1 THEN 1 ELSE 0 END) as low
                FROM moderation_violations
                WHERE created_at > ?
            ", [time() - (7 * 86400)]); // Last 7 days

            echo json_encode([
                'success' => true,
                'stats' => $stats
            ], JSON_PRETTY_PRINT);
            break;

        case 'take_action':
            // Take action on a violation
            $violationId = $_POST['violation_id'] ?? null;
            $actionType = $_POST['action_type'] ?? null;
            $adminUser = $_POST['admin_user'] ?? 'admin';

            if (!$violationId || !$actionType) {
                throw new Exception('Missing required parameters');
            }

            // Get violation details
            $violation = $db->queryOne(
                "SELECT * FROM moderation_violations WHERE id = ?",
                [$violationId]
            );

            if (!$violation) {
                throw new Exception('Violation not found');
            }

            // Perform action
            switch ($actionType) {
                case 'delete':
                    $db->execute(
                        "UPDATE ads SET status = 'inactive' WHERE ad_id = ?",
                        [$violation['ad_id']]
                    );
                    break;

                case 'ban':
                    $db->execute(
                        "UPDATE companies SET status = 'inactive' WHERE company_slug = ?",
                        [$violation['company_slug']]
                    );
                    $db->execute(
                        "UPDATE ads SET status = 'inactive' WHERE company_slug = ?",
                        [$violation['company_slug']]
                    );
                    break;

                case 'approve':
                    // Mark as resolved without action
                    break;

                case 'pause':
                    $db->execute(
                        "UPDATE ads SET status = 'inactive' WHERE ad_id = ?",
                        [$violation['ad_id']]
                    );
                    break;
            }

            // Update violation status
            $db->execute("
                UPDATE moderation_violations
                SET status = 'resolved',
                    resolved_at = ?,
                    resolved_by = ?,
                    action_taken = ?
                WHERE id = ?
            ", [time(), $adminUser, $actionType, $violationId]);

            // Log action
            $db->execute("
                INSERT INTO moderation_actions
                (violation_id, ad_id, action_type, admin_user, reason, created_at)
                VALUES (?, ?, ?, ?, ?, ?)
            ", [
                $violationId,
                $violation['ad_id'],
                $actionType,
                $adminUser,
                $_POST['reason'] ?? 'Admin action',
                time()
            ]);

            // Send notification email to ad owner
            try {
                $notificationSent = $notifier->notifyAdOwner($violation, $actionType, $adminUser);
                $notificationStatus = $notificationSent ? 'sent' : 'failed';
            } catch (Exception $e) {
                error_log("Notification error: " . $e->getMessage());
                $notificationStatus = 'error';
            }

            echo json_encode([
                'success' => true,
                'message' => "Action '$actionType' executed successfully",
                'notification' => $notificationStatus
            ], JSON_PRETTY_PRINT);
            break;

        default:
            throw new Exception('Invalid action');
    }

} catch (Exception $e) {
    http_response_code(500);
    echo json_encode([
        'success' => false,
        'error' => $e->getMessage()
    ], JSON_PRETTY_PRINT);
}

File: python_system/python_shared/api/moderation_violations.py
"""
moderation_violations.py - Moderation Violations API
Get ads flagged by the moderation system
Converted from PHP to Python
"""

from fastapi import APIRouter, Query, Depends
from fastapi.responses import JSONResponse
from sqlalchemy.orm import Session
from pathlib import Path
import json
import time

import sys
sys.path.insert(0, str(Path(__file__).parent.parent))

from models import Ad
from database import get_db

router = APIRouter()

VIOLATIONS_PATH = Path(__file__).parent.parent / "data" / "moderation_violations"
VIOLATIONS_PATH.mkdir(parents=True, exist_ok=True)


@router.get("/moderation_violations")
async def get_violations(
    status: str = Query(None, description="Filter by status: pending, reviewed, resolved"),
    severity: str = Query(None, description="Filter by severity: low, medium, high, critical"),
    limit: int = Query(50, ge=1, le=200),
    db: Session = Depends(get_db)
):
    """Get moderation violations"""
    try:
        violations = []

        # Load violations from files
        for file_path in VIOLATIONS_PATH.glob("*.json"):
            try:
                with open(file_path, "r") as f:
                    data = json.load(f)
                    if isinstance(data, list):
                        violations.extend(data)
                    else:
                        violations.append(data)
            except Exception:
                continue

        # Filter by status
        if status:
            violations = [v for v in violations if v.get("status") == status]

        # Filter by severity
        if severity:
            violations = [v for v in violations if v.get("severity") == severity]

        # Sort by timestamp (newest first)
        violations.sort(key=lambda x: x.get("timestamp", 0), reverse=True)

        # Limit
        violations = violations[:limit]

        # Count by status
        stats = {
            "pending": sum(1 for v in violations if v.get("status") == "pending"),
            "reviewed": sum(1 for v in violations if v.get("status") == "reviewed"),
            "resolved": sum(1 for v in violations if v.get("status") == "resolved")
        }

        return {
            "success": True,
            "violations": violations,
            "stats": stats,
            "total": len(violations)
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )


@router.post("/moderation_violations")
async def create_violation(request):
    """Create a new moderation violation record"""
    try:
        data = await request.json()

        ad_id = data.get("ad_id", "")
        violation_type = data.get("violation_type", "")
        severity = data.get("severity", "medium")
        description = data.get("description", "")
        detected_by = data.get("detected_by", "system")

        if not ad_id or not violation_type:
            return JSONResponse(
                status_code=400,
                content={"success": False, "error": "Missing ad_id or violation_type"}
            )

        # Create violation record
        violation = {
            "id": f"VIO-{int(time.time())}-{ad_id[:8]}",
            "ad_id": ad_id,
            "violation_type": violation_type,
            "severity": severity,
            "description": description,
            "detected_by": detected_by,
            "status": "pending",
            "timestamp": int(time.time()),
            "created_at": time.strftime("%Y-%m-%d %H:%M:%S")
        }

        # Save to daily file
        today = time.strftime("%Y-%m-%d")
        file_path = VIOLATIONS_PATH / f"{today}.json"

        daily_violations = []
        if file_path.exists():
            with open(file_path, "r") as f:
                daily_violations = json.load(f)

        daily_violations.append(violation)

        with open(file_path, "w") as f:
            json.dump(daily_violations, f, indent=2)

        return {
            "success": True,
            "message": "Violation recorded",
            "violation_id": violation["id"]
        }

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )


@router.post("/moderation_violations/{violation_id}/resolve")
async def resolve_violation(violation_id: str, request):
    """Resolve a moderation violation"""
    try:
        data = await request.json()
        action = data.get("action", "")  # approve, block, delete
        notes = data.get("notes", "")

        # Find and update violation
        for file_path in VIOLATIONS_PATH.glob("*.json"):
            try:
                with open(file_path, "r") as f:
                    violations = json.load(f)

                for v in violations:
                    if v.get("id") == violation_id:
                        v["status"] = "resolved"
                        v["resolution_action"] = action
                        v["resolution_notes"] = notes
                        v["resolved_at"] = time.strftime("%Y-%m-%d %H:%M:%S")

                        with open(file_path, "w") as f:
                            json.dump(violations, f, indent=2)

                        return {
                            "success": True,
                            "message": "Violation resolved",
                            "violation_id": violation_id,
                            "action": action
                        }

            except Exception:
                continue

        return JSONResponse(
            status_code=404,
            content={"success": False, "error": "Violation not found"}
        )

    except Exception as e:
        return JSONResponse(
            status_code=500,
            content={"success": False, "error": str(e)}
        )

